### ISSUE TITLE
ARM64 support

### ISSUE THREAD

**Issue by lag-linaro** [neo4j/docker-neo4j#140]
After successfully building/running Neo4J on an ARM64 platform, there doesn't seem to be any reason for not adding it as a supported architecture in Docker Hub.

Is this something you guys have thought about?

**Comment by lag-linaro** [comment]
Just to keep you guys in the loop, I've taken the initiative to place a fork of this repo into ARM64 CI.  In this particular instance CodeFresh has been chosen, but there are other ARM64 based CI platforms freely available (Shippable, WorksOnARM, etc).

When using CodeFresh you can choose to supply pull, build, test, push instructions either inline (stored privately in CodeFresh) or via a YMAL file in the repo.  For clarity and openness it currently resides in the forked code base.

[YAML file](https://github.com/lag-linaro/docker-neo4j/blob/community-3.4.6/codefresh.yml)

As you'll probably notice by reading the script, a built image is pushed to Docker Hub for testing/consumption.  It can be pulled with:

`docker pull laglinaro/neo4j:3.{3|4}.6-{community|enterprise}`

Where the tag (post ':') can be any one of the following:

```
3.3.6-community
3.3.6-enterprise
3.4.6-community
3.4.6-enterprise
```

Hopefully this goes some way to show that build testing (you can also run/launch test) on AArch64 (ARM64) isn't difficult, despite not owning actual H/W.

**Comment by lag-linaro** [comment]
Have you guys had any more thoughts on this?

Lots of available hardware/services for `aarch64` now.  Even Amazon (AWS) EC2 instances.

**Comment by motion-x86** [comment]
Second that, ARM64 containers would be great. Any updates on this thread?

**Comment by andresvidal** [comment]
Second that second. ARM64 is everywhere. I checked docker hub and nothing yet.

**Comment by MrRobJohnson** [comment]
Second that second second.

Not sure why as there are plenty of examples of getting neo4j installed on the Raspberry Pi but no official Docker image for it.

Lag Linaro image works a treat.  Thanks 👍

**Comment by kahekahe** [comment]
Is `laglinaro/neo4j:3.3.6-community` working on raspberry pi 4? 

I tried to run it, but got

```
neo4j_1  | standard_init_linux.go:211: exec user process caused "exec format error"

neo4j_neo4j_1 exited with code 1
```

**Comment by lag-linaro** [comment]
That branch is now old and probably suffers with bitrot.

Please use the parent project instead.

**Comment by kahekahe** [comment]
As in the [official neo4j docker image](<https://hub.docker.com/_/neo4j>)?

It doesn't look like it supports raspberry pi

```
Using default tag: latest
latest: Pulling from library/neo4j
no matching manifest for linux/arm/v7 in the manifest list entries
```

or do you mean something else?

**Comment by kahekahe** [comment]
It is a raspberry pi 4, however `cat /proc/cpuinfo` returns `model name	: ARMv7 Processor rev 3 (v7l)`

Looks like neo4j only supports `linux/amd64`

```shell
pi@rpi4:~/Projects/neo4j $ sudo docker run --rm mplatform/mquery  neo4j
Image: neo4j
 * Manifest List: Yes
 * Supported platforms:
   - linux/amd64

```

**Comment by lag-linaro** [comment]
That's right, which is what this issue is about - providing support for ARM64.

However, if you pull the repo and build it yourself, it'll run.

**Comment by kahekahe** [comment]
Ok, thanks. 

Doesn't look like it works anymore, I'm getting a similar error when I try to build from the repo



```shell
Sending build context to Docker daemon  153.9MB
Step 1/13 : FROM openjdk:8-jre-slim
8-jre-slim: Pulling from library/openjdk
no matching manifest for linux/arm/v7 in the manifest list entries
make: *** [Makefile:55: tmp/.image-id-community] Error 1
```

**Comment by lag-linaro** [comment]
You're not on an Armv7 platform. Until you fix that, it's a non-starter.

FYI, I haven't tested Armv7 only Armv8 (AArch64, Arm64), so can't vouch for it.

The "exec format error" happens when you try to execute a binary that does not match your system architecture. Armv7 binary on an Armv8 platform in your case.

**Comment by khalidtn2** [comment]
any updates on this thread? Is this possible to run? if so, are there any steps to enable that?

**Comment by harshilishere** [comment]
Can somebody please guide me on how can i build the container in someway? So I can also make the possible contribution. I have rpi 4 4gb ram as of now. Please tell me what can I do to make arm supported container for Neo4j.

**Comment by harshilishere** [comment]
I have found this:
https://gist.github.com/wilsongp/2088c40a56d776d440ef1511075206c4

**Comment by zeddD1abl0** [comment]
@harshilishere  The problem you're encountering is that your RPi4 is running probably Raspbian, which is 32bit only. You need to use something like Ubuntu Server 64bit for the image provided by @lag-linaro to work as expected. I don't think armhp or arm32 will ever actually work with Neo4j.

**Comment by harshilishere** [comment]
@zeddD1abl0 thanks for replying. yeah i know it will work on amd64 as of now. But that's what i am asking. How can i compile one and give it to others to test!? And maybe by doing this we can use neo4j on small projects too?!

**Comment by folterj** [comment]
I take it there is still no ARM support?
We would be interested to run on Amazon AWS EC2 Arm (A1 etc) instances - the majority of Docker images currently supports Arm architecture

**Comment by jennyowen** [comment]
The reason we haven't provided arm64 compatible docker images is that arm64 isn't a supported architecture for Neo4j.

It may well be possible to run Neo4j server on arm64 machines, but that's not something we test in any capacity. We couldn't provide any type of assurance that an arm64 Neo4j image would work. Even though it is be straightforward to *create* a Neo4j arm64 image, it wouldn't be properly tested, and we definitely wouldn't meet the quality requirements for official docker images.

HOWEVER!

Although I absolutely can't create an official Neo4j arm64 image at https://hub.docker.com/_/neo4j we do have an experimental docker repository at `neo4j/neo4j-experimental` which we mostly use for pre-releases: 
https://hub.docker.com/r/neo4j/neo4j-experimental

The images there are unsupported and NOT ok for production.
I could publish arm64 images there on the understanding that they are **unsupported**.
I've created pull request #265 for this issue.

**Comment by folterj** [comment]
Building neo4j for Arm is actually really easy.

Requirements
- Docker
- Enable experimental features (Google if you don't know how to do this)

Copy files from desired neo4j version (https://hub.docker.com/_/neo4j):
- Dockerfile
- local-package\neo4jlabs-plugins.json

Important! Enabling tini for Arm:
- In the neo4j Dockerfile contents, replace 'tini' with 'tini-arm64' and fix or remove the SHA256 check, e.g.:
> curl -L --fail --silent --show-error "https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini-arm64" > /sbin/tini

Build & push with buildx (note that due to architecture-specific tini this image will only work on Arm64):
> docker buildx build --platform linux/amd64,linux/arm64 --push -t [imagename] .

You can now load and run this image on AWS Arm after installing docker (e.g. amazon-linux-extras install docker)

I don't understand why Arm wouldn't be supported (or isn't a supported architecture for Neo4j), as there is no doubt a large interest running images on Amazon AWS Arm architecture.

**Comment by jennyowen** [comment]
@folterj oh no, it looks as though you didn't bother to read what I said before assuming I don't know how to do my job.
I will repeat myself with smaller words so that maybe you will trouble yourself to read all the way through.

1.  Neo4j is not supported on ARM64. Not just the docker image but the core product. Maybe we will add support in the future (it's being considered) but for now that's just how it is.
1. Neo4j, all of it, is untested on ARM64.
   - We have all sorts of tests for neo4j, especially for things like data integrity and performance. This is something that a different architecture might break. There is a lot more to testing a database than just making sure it starts.
3. Creating a Neo4j arm64 docker image is obviously easy.
4. Testing a Neo4j arm64 docker image is hard. 
   - Remember, we don't test the *core* product on arm64, so there is no guarantee a docker image will work properly.
   - In software engineering, it's generally considered a good idea to test stuff before releasing it officially.
5. **(This point is important!)** I already made changes to the make scripts so that anyone can build an arm64 compatible image: #265
6. **(This point is the most important!)** I already said I was going to release arm64 images on our unofficial dockerhub repo.

**Comment by folterj** [comment]
@jennyowen ok thank you for clarifying things now, including clarifications to your original post. Apologies, I did not see this entire message when I originally read read it. I was certainly not implying lack of skill or effort, merely surprised by the formal statement regarding no Arm support - reading this as there was no interest in supporting this at all. I hope this will change in the future (perhaps a bit like Java 11).
Interestingly we've tested neo4j with a database on Arm and found no issues at all.

**Comment by jennyowen** [comment]
I've created a new repository in dockerhub for arm64 images here:
https://hub.docker.com/r/neo4j/neo4j-arm64-experimental
and updated the readme about it:
https://github.com/neo4j/docker-neo4j#neo4j-images-for-arm64

please try these images if you need Neo4j on arm64.

**Comment by MarianAlexandruAlecu** [comment]
```
> NEO4JVERSION=3.5.28 make -f make-arm64.mk clean tag-arm
> docker run --publish=7474:7474 --publish=7687:7687 --volume=$HOME/db/data:/data neo4j/neo4j-arm64-experimental:3.5.28-arm64
```

Doesn't work on Apple Silicon M1. I get:
```
/lib64/ld-linux-x86-64.so.2: No such file or directory
```

Is there an obvious fix that I'm missing?

**Comment by MarianAlexandruAlecu** [comment]
I've tried to replace the base image with this `arm64v8/openjdk:8-jre-alpine` in the [official 3.5.5 image](https://github.com/neo4j/docker-neo4j-publish/blob/master/3.2.1/community/Dockerfile)

But when I'm starting this image I'm getting
```
Exception in thread "main" java.lang.UnsatisfiedLinkError: no zstd-jni in java.library.path
Unsupported OS/arch, cannot find /linux/aarch64/libzstd-jni.so or load zstd-jni from system libraries. Please try building from source the jar or providing libzstd-jni in you system.
```
Looks like it's related to this issue https://github.com/luben/zstd-jni/issues/43

Any solutions to make 3.5.5 run on M1/Docker?

**Comment by MarianAlexandruAlecu** [comment]
If someone wants to run `Neo4J 3.5.28` on Apple Silicon M1 + Docker do this
1. Clone just this folder https://github.com/neo4j/docker-neo4j-publish/tree/master/3.5.28/community
2. Update the Dockerfile with this:
```
FROM arm64v8/openjdk:8-jdk

ENV NEO4J_SHA256=d08415b80d4d6956ccd6063668b6d14d31905c41a712da1522b73079e7dc0aa1 \
    NEO4J_TARBALL=neo4j-community-3.5.28-unix.tar.gz \
    NEO4J_EDITION=community \
    NEO4J_HOME="/var/lib/neo4j" \
    TINI_VERSION="v0.18.0" \
    TINI_SHA256="7c5463f55393985ee22357d976758aaaecd08defb3c5294d353732018169b019"
ARG NEO4J_URI=https://dist.neo4j.org/neo4j-community-3.5.28-unix.tar.gz

RUN addgroup --system neo4j && adduser --system --no-create-home --home "${NEO4J_HOME}" --ingroup neo4j neo4j

COPY ./local-package/* /tmp/

RUN apt update \
     && apt install -y curl wget gosu jq \
     && curl -L --fail --silent --show-error "https://github.com/krallin/tini/releases/download/${TINI_VERSION}/tini-arm64" > /sbin/tini \
     && echo "${TINI_SHA256}  /sbin/tini" | sha256sum -c --strict --quiet \
     && chmod +x /sbin/tini \
     && curl --fail --silent --show-error --location --remote-name ${NEO4J_URI} \
     && echo "${NEO4J_SHA256}  ${NEO4J_TARBALL}" | sha256sum -c --strict --quiet \
     && tar --extract --file ${NEO4J_TARBALL} --directory /var/lib \
     && mv /var/lib/neo4j-* "${NEO4J_HOME}" \
     && rm ${NEO4J_TARBALL} \
     && mv "${NEO4J_HOME}"/data /data \
     && mv "${NEO4J_HOME}"/logs /logs \
     && chown -R neo4j:neo4j /data \
     && chmod -R 777 /data \
     && chown -R neo4j:neo4j /logs \
     && chmod -R 777 /logs \
     && chown -R neo4j:neo4j "${NEO4J_HOME}" \
     && chmod -R 777 "${NEO4J_HOME}" \
     && ln -s /data "${NEO4J_HOME}"/data \
     && ln -s /logs "${NEO4J_HOME}"/logs \
     && mv /tmp/neo4jlabs-plugins.json /neo4jlabs-plugins.json \
     && rm -rf /tmp/* \
     && rm -rf /var/lib/apt/lists/* \
     && apt-get -y purge --auto-remove curl

ENV PATH "${NEO4J_HOME}"/bin:$PATH

WORKDIR "${NEO4J_HOME}"

VOLUME /data /logs

COPY docker-entrypoint.sh /docker-entrypoint.sh

EXPOSE 7474 7473 7687

ENTRYPOINT ["/sbin/tini", "-g", "--", "/docker-entrypoint.sh"]
CMD ["neo4j"]
```

I changed:
- The base image. I was having issues with JRE so I went full JDK
- Move to Tini arm64

P.S. I explicitly wrote the name of the processor for people to hopefully find it when searching online and to save time

**Comment by rappster** [comment]
@jennyowen I tried pulling the experimental image, but got this

```
Using default tag: latest
Error response from daemon: manifest for neo4j/neo4j-arm64-experimental:latest not found: manifest unknown: manifest unknown
```

**Comment by jennyowen** [comment]
@rappster We don't push a `latest` tag to there, it's a pretty informal repository.
You will need to specify the neo4j version you want when pulling the tag. eg `neo4j/neo4j-arm64-experimental:4.3.4`.

**Comment by johannb75** [comment]
Hello, I recently moved to mac with the m1 chip.
I tried installing the experimental docker image of the enterprise version supporting arm64.
For my projects, I need to use `apoc`.
When I run this command to start neo4j:
```bash
docker run \
-p 7474:7474 -p 7687:7687 \
--name neo4j \
-e NEO4J_apoc_export_file_enabled=true \
-e NEO4J_apoc_import_file_enabled=true \
-e NEO4J_apoc_import_file_use__neo4j__config=true \
-e NEO4JLABS_PLUGINS=\[\"apoc\"\] \
-e NEO4J_ACCEPT_LICENSE_AGREEMENT=yes \
neo4j/neo4j-arm64-experimental:4.4.0-alpha01-enterprise
```
I get this error:
```bash
Fetching versions.json for Plugin 'apoc' from https://neo4j-contrib.github.io/neo4j-apoc-procedures/versions.json
Error: No jar URL found for version '4.4.0-alpha01' in versions.json from 'https://neo4j-contrib.github.io/neo4j-apoc-procedures/versions.json'
```
Followed by a huge JSON:
```json
[
    {
        "neo4j": "4.3.6",
        "neo4jVersion": "4.3.6",
        "apoc": "4.3.0.3",
        "version": "4.3.0.3",
        "url": "http://github.com/neo4j-contrib/neo4j-apoc-procedures/releases/4.3.0.3",
        "homepageUrl": "http://github.com/neo4j-contrib/neo4j-apoc-procedures/releases/4.3.0.3",
        "jar": "https://github.com/neo4j-contrib/neo4j-apoc-procedures/releases/download/4.3.0.3/apoc-4.3.0.3-all.jar",
        "downloadUrl": "https://github.com/neo4j-contrib/neo4j-apoc-procedures/releases/download/4.3.0.3/apoc-4.3.0.3-all.jar",
        "sha1": "def2825a180727048829004f9a85c7cafa895480",
        "sha256": "656ac1afca99766bff832acf6da229d086d8a721e4f1d2739dc28417c156dab6",
        "md5": "d26f201c4d9c120a2c239d282a335ad6",
        "config": {
            "+:dbms.security.procedures.unrestricted": [
                "apoc.*"
            ]
        }
    },
    {
        "neo4j": "4.3.5",
        "neo4jVersion": "4.3.5",
        "apoc": "4.3.0.3",
        "version": "4.3.0.3",
        "url": "http://github.com/neo4j-contrib/neo4j-apoc-procedures/releases/4.3.0.3",
        "homepageUrl": "http://github.com/neo4j-contrib/neo4j-apoc-procedures/releases/4.3.0.3",
        "jar": "https://github.com/neo4j-contrib/neo4j-apoc-procedures/releases/download/4.3.0.3/apoc-4.3.0.3-all.jar",
        "downloadUrl": "https://github.com/neo4j-contrib/neo4j-apoc-procedures/releases/download/4.3.0.3/apoc-4.3.0.3-all.jar",
        "sha1": "def2825a180727048829004f9a85c7cafa895480",
        "sha256": "656ac1afca99766bff832acf6da229d086d8a721e4f1d2739dc28417c156dab6",
        "md5": "d26f201c4d9c120a2c239d282a335ad6",
        "config": {
            "+:dbms.security.procedures.unrestricted": [
                "apoc.*"
            ]
        }
    },
    // goes on and on like this......
]
```

I honestly don't know what to do with this.

**Comment by rweverwijk** [comment]
Hi @johannb75,

I'm not part of the Neo4j team, but my suggestion is

1. download the APOC jar manually, place it in a directory and mount that directory to `/plugins`
2. remove the NEO4JLABS_PLUGINS environment variable

After that it should work.

Cheers,
Ron

**Comment by jennyowen** [comment]
@johannb75 yeah the `NEO4JLABS_PLUGINS` doesn't have the best error behaviour. 
The reason APOC is failing is because Neo4j 4.4.0 is not released yet, it's only in alpha at the moment (8th Nov 2021). There isn't an APOC plugin for it just yet, so the docker image is complaining because it can't download APOC for you.

Normally, @rweverwijk 's solution would be a good one, but since there is currently no 4.4 apoc your options are:

- use a 4.3 image instead (there are no enterprise 4.3 images though because they're untested on arm and unsuitable for production)
- use `NEO4JLABS_PLUGINS=\[\"apoc-core\"\]` instead, and do without the extended apoc functionality until 4.4 is fully released.

**Comment by jennyowen** [comment]
For future reference of anyone reaching here through Google, Neo4j 4.4 onwards officially supports ARM64 architecture.

The official neo4j images you get through dockerhub will work with an ARM64 machine, in exactly the same way as a regular amd64 architecture:

For example: 
`docker run   --publish=7474:7474 --publish=7687:7687 neo4j:4.4.0`

The official dockerhub page for our arm64 image is;
https://hub.docker.com/r/arm64v8/neo4j

**Comment by MarianAlexandruAlecu** [comment]
@jennyowen Is there any change to get ARM support for `neo4j:3.x.y`?

Specific case: dev using M1 Mac on 3 years old project

**Comment by jennyowen** [comment]
@MarianAlexandruAlecu ok I pushed an ARM image for 3.5.30: `neo4j/neo4j-arm64-experimental:3.5.30`. I also updated the release automation so that any new 3.5 releases will get an arm image. Hopefully that makes things easier for you.

As for getting ARM support, there are no plans to support ARM on any Neo4j version before 4.4. 
The images in `neo4j/neo4j-arm64-experimental` are not supported and I only provided them because so many people were asking. They're on a "use at your own risk" kinda deal.

In general, 3.5 is the only 3.x version we currently support.
